Support dynamically linking against system double-conversion library

musicinmybrain opened the pull request:
New environment variables `UJSON_BUILD_DC_INCLUDES` and `UJSON_BUILD_DC_LIBS` allow overriding the include path for double-conversion and adding linker flags for an external double-conversion library. They should generally be used together.

This is particularly useful to Linux distribution packagers who would otherwise need a patch to unbundle `double-conversion`. (It may still be necessary to `rm -rf deps/double-conversion`; I have not checked, as Fedora Linux would require this anyway.)

Fixes #375

Changes proposed in this pull request:

* Allow overriding default include path/paths for `double-conversion` with environment variable `UJSON_BUILD_DC_INCLUDES`. This allows multiple whitespace-delimited paths but as a consequence doesn’t support paths containing whitespace, which is probably acceptable for the intended uses. For example, Linux distributions building against a system copy of `double-conversion` will want something like `UJSON_BUILD_DC_INCLUDES=/usr/include/double-conversion`.
* Allow adding linker flags for `double-conversion` with environment variable `UJSON_BUILD_DC_LIBS`; by default there are still none. For example, Linux distributions building against a system copy of `double-conversion` will want something like `UJSON_BUILD_DC_LIBS=-ldouble-conversion`.

musicinmybrain committed:
Add env vars to build with system double-conversion

This fixes #376 and is useful to Linux distribution packagers.

New environment variables UJSON_BUILD_DC_INCLUDES and UJSON_BUILD_DC_LIBS allow overriding the include path for double-conversion and adding linker flags for an external double-conversion library. They should generally be used together.

bwoodsend commented:
Could you add a little *Build Options* section to the end of the readme to illustrate to others how to link against a system `double-conversion`?

musicinmybrain commented:
Yes. I can add that to this PR.

bwoodsend commented on setup.py:
```diff
@@ -25,9 +25,15 @@
     extra_compile_args=["-D_GNU_SOURCE"],
-    extra_link_args=["-lstdc++", "-lm"] + strip_flags,
+    extra_link_args=["-lstdc++", "-lm"]
+    + environ.get("UJSON_BUILD_DC_LIBS", "").split()
```
Normally this kind of thing is split by `os.pathsep`. I'd be very surprised if a path with spaces ever did need to be added there. With `shlex.split()` would at least let you quote or backslash escape a space if need be.

musicinmybrain replied:
Thanks for the suggestions.

I tend to think that the `shlex.split()` approach is appropriate for `UJSON_BUILD_DC_LIBS`, since `LDFLAGS`-style environment variables are usually like `-L/usr/lib/foo -lbar -lbat`. There’s some risk of surprising magic by doing this, but it pretty closely resembles the behavior of traditional build systems that end up doing something like `${CC} ${CFLAGS} ${LDFLAGS} …` with unquoted shell interpolation.

```python
shlex.split(environ.get("UJSON_BUILD_DC_LIBS", ""))
```

Meanwhile, the proposed `UJSON_BUILD_DC_INCLUDES` is a list-of-paths like `PATH` rather than a set of `CFLAGS`-style compiler flags, so splitting on `os.pathsep`, i.e. `:`, does seem reasonable there.

```python
environ.get("UJSON_BUILD_DC_INCLUDES", "./deps/double-conversion/double-conversion").split(pathsep)
```

Sound good?

musicinmybrain committed:
Improved splitting for new env vars

musicinmybrain committed:
Document build options in README.md

musicinmybrain committed:
External dconv improvement

Automatically disable building ./deps/double-conversion when UJSON_BUILD_DC_LIBS is set to a non-empty value.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

musicinmybrain commented:
Made some changes based on feedback. Ensured the bundled `double-conversion` is not built if `UJSON_BUILD_DC_LIBS` is set to a non-empty value. All further feedback is welcome.

bwoodsend commented on setup.py:
```diff
@@ -1,10 +1,18 @@
+dconv_includes = environ.get(
+    "UJSON_BUILD_DC_INCLUDES",
+    "./deps/double-conversion/double-conversion",
+).split(pathsep)
```
This would be `[""]` if `UJSON_BUILD_DC_INCLUDES` is not set. I'm not sure if it's likely to do any damage but I'd rather the current working directory isn't added to include path. Can you make sure that empty strings are removed.

musicinmybrain replied:
> This would be `[""]` if `UJSON_BUILD_DC_INCLUDES` is not set.

To be pedantic, it would be `["./deps/double-conversion/double-conversion"]` if the variable were not set, and `[""]` only if the variable were set to the empty string.

I can easily filter out empty strings, though; anyone who wants to add the current working directory as a relative path for some reason could still use `.`.

musicinmybrain committed:
Filter out any empty strings from dconv_includes

bwoodsend commented:
Do you need ujson to release this before you can go back to your work on the Fedora RPM?

musicinmybrain commented:
> Do you need ujson to release this before you can go back to your work on the Fedora RPM?

Not really—we can easily carry it as a downstream patch for now.
